File: composeApp/src/androidMain/kotlin/com/jetbrains/kmpapp/screens/AuctionScreen.kt
package com.jetbrains.kmpapp.screens

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonColors
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import coil3.compose.rememberAsyncImagePainter
import com.jetbrains.kmpapp.component.TopBar
import com.jetbrains.kmpapp.viewmodel.AuctionScreenState
import com.jetbrains.kmpapp.viewmodel.AuctionViewModel
import org.koin.androidx.compose.koinViewModel

@Composable
fun AuctionScreen(
    auctionId: Int,
    onBack: () -> Unit,
) {
    val viewModel: AuctionViewModel = koinViewModel()
    val screenState by viewModel.state.collectAsStateWithLifecycle()

    var bidDialogVisible by remember { mutableStateOf(false) }
    var bidInput by remember { mutableStateOf("") }

    LaunchedEffect(auctionId) {
        viewModel.initialize(auctionId)
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(8.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {

        // content
        when (val state = screenState) {
            is AuctionScreenState.Loading -> {
                Text("Loading data ...")
            }

            is AuctionScreenState.Error -> {
                Text(state.message)
            }

            is AuctionScreenState.Details -> {
                val auction = state.auction
                // top bar
                TopBar(
                    title = auction.product.name,
                    onBack = onBack
                )
                if (bidDialogVisible) {
                    AlertDialog(
                        onDismissRequest = { bidDialogVisible = false },
                        title = { Text("Enter your bid") },
                        text = {
                            Column {
                                Text("Enter a higher bid than the current:")
                                TextField(
                                    value = bidInput,
                                    onValueChange = { bidInput = it },
                                    singleLine = true
                                )
                            }
                        },
                        confirmButton = {
                            TextButton(onClick = {
                                val bid = bidInput.toIntOrNull()
                                if (bid != null && bid > auction.product.currentBid) {
                                    // save in memory
                                    viewModel.placeNewBid(bid)
                                    bidDialogVisible = false
                                }
                            }) {
                                Text("Submit")
                            }
                        },
                        dismissButton = {
                            TextButton(onClick = { bidDialogVisible = false }) {
                                Text("Cancel")
                            }
                        }
                    )
                }

                // image
                ProductLargeImage(
                    modifier = Modifier.weight(1f),
                    url = auction.product.mainImage?.imageUrlLarge
                )

                // name
                Text(
                    text = auction.product.name,
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                )

                // description
                Text(
                    text = auction.product.description ?: "",
                    style = MaterialTheme.typography.bodySmall,
                    maxLines = 2,
                )

                // municipality name
                Text(
                    text = auction.product.municipalityName,
                    style = MaterialTheme.typography.bodySmall,
                    maxLines = 1,
                )

                // current bid
                Text(
                    text = "Current bid: ${auction.product.currentBid}",
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                )

                // categories
                Column(
                    verticalArrangement = Arrangement.spacedBy(4.dp),
                    horizontalAlignment = Alignment.Start,
                ) {
                    // category 1
                    auction.category1?.let {
                        Text(
                            text = it.headline,
                            color = Color.Blue,
                        )
                    }

                    // category 2
                    auction.category2?.let {
                        Text(
                            text = it.headline,
                            color = Color.Blue,
                        )
                    }

                    // category 3
                    auction.category3?.let {
                        Text(
                            text = it.headline,
                            color = Color.Blue,
                        )
                    }
                }

                // end date
                Text(
                    text = auction.product.endDate,
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                )

                // place bid button
                Button(
                    onClick = { bidDialogVisible = true },
                    colors = ButtonColors(
                        containerColor = Color.Red,
                        contentColor = Color.White,
                        disabledContainerColor = Color.Red,
                        disabledContentColor = Color.White,
                    ),
                    shape = RoundedCornerShape(8.dp)
                ) {
                    Text("Place bid")
                }
            }
        }
    }
}

@Composable
fun ProductLargeImage(
    modifier: Modifier = Modifier,
    url: String?,
) {
    if (url != null) {
        Image(
            modifier = modifier
                .fillMaxSize()
                .clip(RoundedCornerShape(8.dp)),
            painter = rememberAsyncImagePainter(model = url),
            contentDescription = "product thumbnail",
            contentScale = ContentScale.Crop,
        )
    } else {
        Image(
            imageVector = Icons.Default.Lock,
            contentDescription = "product placeholder",
            modifier = modifier
                .fillMaxSize()
                .clip(RoundedCornerShape(8.dp)),
            contentScale = ContentScale.Crop,
        )
    }
}

File: iosApp/iosApp/Screen/AuctionScreen.swift
import SwiftUI
import Shared
import KMPObservableViewModelSwiftUI

struct AuctionScreen: View {
    let auctionId: Int32

    @StateViewModel
    var viewModel = AuctionViewModel(repository: KoinDependencies().auctionsRepository)

    @State private var bidDialogVisible = false
    @State private var bidInput = ""

    var body: some View {
        VStack(spacing: 16) {
            switch viewModel.state {
            case is AuctionScreenState.Loading:
                Text("Loading data...")

            case let details as AuctionScreenState.Details:
                let auction = details.auction

                TopBar(title: auction.product.name)

                if bidDialogVisible {
                    BidInputDialog(
                        currentBid: auction.product.currentBid,
                        bidInput: $bidInput,
                        onSubmit: {
                            if let bid = Int(bidInput), bid > auction.product.currentBid {
                                viewModel.placeNewBid(bid: Int32(bid))
                                bidDialogVisible = false
                            }
                        },
                        onCancel: {
                            bidDialogVisible = false
                        }
                    )
                }

                ProductLargeImage(url: auction.product.mainImage?.imageUrlLarge)

                Text(auction.product.name)
                    .font(.body)
                    .fontWeight(.semibold)
                    .lineLimit(1)

                Text(auction.product.description_ ?? "")
                    .font(.footnote)
                    .lineLimit(2)

                Text(auction.product.municipalityName)
                    .font(.footnote)

                Text("Current bid: \(auction.product.currentBid)")
                    .font(.body)
                    .fontWeight(.semibold)

                VStack(alignment: .leading, spacing: 4) {
                    ForEach([auction.category1, auction.category2, auction.category3].compactMap { $0 }, id: \.id) { category in
                        Text(category.headline)
                            .foregroundColor(.blue)
                    }
                }

                Text(auction.product.endDate)
                    .font(.footnote)
                    .fontWeight(.semibold)

                Button("Place bid") {
                    bidDialogVisible = true
                }
                .buttonStyle(.borderedProminent)
                .tint(.red)
                .clipShape(RoundedRectangle(cornerRadius: 8))

            case let error as AuctionScreenState.Error:
                Text(error.message)
                    .foregroundColor(.red)

            default:
                EmptyView()
            }
        }
        .padding(12)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
        .onAppear {
            viewModel.initialize(id: auctionId)
        }
    }
}

struct BidInputDialog: View {
    let currentBid: Int32
    @Binding var bidInput: String
    var onSubmit: () -> Void
    var onCancel: () -> Void

    var body: some View {
        VStack(spacing: 16) {
            Text("Enter your bid")
                .font(.headline)

            Text("Enter a higher bid than the current:")
                .font(.subheadline)

            TextField("Bid amount", text: $bidInput)
                .keyboardType(.numberPad)
                .textFieldStyle(.roundedBorder)

            HStack {
                Button("Cancel", action: onCancel)
                Spacer()
                Button("Submit", action: onSubmit)
            }
        }
        .padding()
        .background(.ultraThinMaterial)
        .clipShape(RoundedRectangle(cornerRadius: 12))
        .shadow(radius: 5)
    }
}


struct ProductLargeImage: View {
    var url: String?

    var body: some View {
        ZStack {
            Color.gray.opacity(0.2)
                .frame(height: 200)
                .clipShape(RoundedRectangle(cornerRadius: 8))
            Group {
                if let url = url, let imageURL = URL(string: url) {
                    AsyncImage(url: imageURL) { image in
                        image.resizable()
                    } placeholder: {
                        ProgressView()
                    }
                } else {
                    EmptyView()
                }
            }
            .frame(height: 200)
            .clipShape(RoundedRectangle(cornerRadius: 8))
        }
    }
}

